### PR TITLE
feat!: new `IConsensus` interface

### LINKED ISSUE #152: Improve `IConsensus` for node developers
## 📚 Context

Currently, the node only knows how to control `Authority` contracts, but not any contract that implements the `IConsensus` interface.

Furthermore, the node calls the `getClaim` function to check if a claim has been submitted yet. This method, however, is unstable, because the error message depends on the Ethereum node implementation. 

Finally, the consensus contract does not emit any event when a claim is submitted successfully. This, again, makes it harder for the node to keep track of which claims have been submitted already.

We should, therefore, improve the `IConsensus` interface with node developers in mind.

## ✔️ Solution

After discussing with @gligneul, @pedroargento and @torives, we agreed that the `IConsensus` interface should:

- Have a function to submit claims (to be used by validator nodes)
- Have a function to retrieve claims (to be used on-chain by `ICartesiDApp` contracts)
- Emit an event signalling when a claim is submitted (to be used by validator nodes, so the validator doesn't try to submit a claim again)
- Emit an event signalling when a claim is accepted (to be used by reader and validators nodes to list all claims that can be retrieved on-chain, and therefore don't need to be submitted)

As a result, we may remove the following contracts:
- `IHistory`
- `History`
- `IHistoryFactory`
- `HistoryFactory`
- `IAuthorityHistoryPairFactory`
- `IAuthorityHistoryPairFactory`

We can define a contract that deals with claim storage and make `Authority` (and `Quorum` in the future) inherit it. We can name it `ClaimStorage`. Other names are welcome!

### PR DESCRIPTION
**Pull request by guidanoli**
Closes #152 and #153
FYI @gligneul

### CODE REVIEW

File: onchain/rollups/contracts/library/LibProof.sol [hunk 1]
```diff
@@ -0,0 +1,21 @@
+// (c) Cartesi and individual authors (see AUTHORS)
+// SPDX-License-Identifier: Apache-2.0 (see LICENSE)
+
+pragma solidity ^0.8.8;
+
+import {Proof} from "../common/Proof.sol";
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/library/LibProof.sol:6]: I'd like to decouple those things.

**Comment by guidanoli** [comment]
> I'm thinking whether we can get rid of the current `AbstractConsensus.sol` but replace it with the codes in `EpochHashStorage.sol`. So the inheritance chain becomes simpler: `IConsensus` -> `EpochHashStorage`(can be renamed as `AbstractConsensus`) -> `Authority`. What do you think?

I'd like `EpochHashStorage` to be completely decoupled from `IConsensus`.
With this, `AbstractConsensus` becomes the "glue" that half-implements `IConsensus` using `EpochHashStorage`.

File: onchain/rollups/contracts/consensus/IConsensus.sol [hunk 2]
```diff
@@ -3,74 +3,68 @@
...
-///
+import {InputRange} from "../common/InputRange.sol";
+
+/// @notice Provides epoch hashes for DApps.
+/// @notice An epoch hash is produced after the machine processes a range of inputs and the epoch is finalized.
+/// This hash serves as proof for the outputs produced by the machine during the epoch.
+/// @notice Empty input ranges are not accepted, since at least one input is necessary to advance the state of the machine.
+/// @notice Validators may synchronize epoch finalization, but such mechanism is not specified by this interface.
+/// @notice Regardless of synchronization or the lack thereof,
+/// validators shouldn't need to submit a claim if it was already submitted by them or accepted by the consensus.
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/IConsensus.sol:14]: > Part of this sentence means that validators shouldn't submit if it's already submitted

If you read the whole sentence, it says "submitted by them", and by "them" I mean the validator itself! Take quorum, for example. If a validator has voted for a claim, they shouldn't need to vote for it again later. Their second vote will be completely discarded. Maybe I should rephrase this to make it clearer to the user. :-)

File: onchain/rollups/contracts/consensus/IConsensus.sol [hunk 3]
```diff
@@ -3,74 +3,68 @@
...
+        bytes32 epochHash
+    );
+
+    /// @notice A claim was accepted by the consensus.
+    /// @param dapp The DApp contract address
+    /// @param inputRange The input range
+    /// @param epochHash The epoch hash
+    /// @dev The input range MUST NOT be empty.
+    /// @dev Overwrites any previous `ClaimAcceptance(dapp, inputRange, _)`.
+    /// @dev MUST be triggered whenever `getEpochHash(dapp, inputRange)` changes to `epochHash`.
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/IConsensus.sol:42]: Yeah, but what does it mean for an epoch hash to be accepted, in terms of this interface? My conclusion was that it has to do with the `getEpochHash` function.

File: onchain/rollups/contracts/consensus/IConsensus.sol [hunk 4]
```diff
@@ -3,74 +3,68 @@
...
-        returns (
-            bytes32 epochHash_,
-            uint256 firstInputIndex_,
-            uint256 lastInputIndex_
-        );
+    /// @notice Submit a claim to the consensus.
+    /// @param dapp The DApp contract address
+    /// @param inputRange The input range
+    /// @param epochHash The epoch hash
+    /// @dev MUST trigger a `ClaimSubmission` event.
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/IConsensus.sol:54]: I'm following EIP-20 and EIP-1155, and saying when an event is triggered both in the event doc and in the function doc.

File: onchain/rollups/test/foundry/consensus/authority/Authority.t.sol [hunk 5]
```diff
@@ -65,227 +58,49 @@ contract AuthorityTest is TestBase {
...
+        Authority authority = new Authority(_owner);
 
-        authority = new Authority(_owner);
+        vm.expectEmit(true, true, false, true, address(authority));
+        emit ClaimSubmission(_owner, _dapp, _inputRange, _epochHash);
 
-        vm.prank(_owner);
-        authority.setHistory(historyR);
-        assertEq(address(authority.getHistory()), address(historyR));
+        vm.expectEmit(true, true, false, true, address(authority));
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/test/foundry/consensus/authority/Authority.t.sol:98]: True!

File: onchain/rollups/contracts/library/LibProof.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/library/LibProof.sol:6]: Suppose you use `Proof` but not `LibProof`. If they live on the same file, and I make a change in `LibProof`, even if `Proof` stays the same, this might trigger a redeployment for you, right?

File: onchain/rollups/contracts/dapp/CartesiDApp.sol [hunk 6]
```diff
@@ -199,43 +190,20 @@ contract CartesiDApp is
         bytes calldata _notice,
         Proof calldata _proof
     ) external view override {
-        bytes32 epochHash;
-        uint256 firstInputIndex;
-        uint256 lastInputIndex;
+        // calculate the input index, and check if it is inside the range
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/dapp/CartesiDApp.sol:193]: Even though we are not using the input index, we need to calculate it (internally), right?

File: onchain/rollups/contracts/consensus/IConsensus.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/IConsensus.sol:54]: I decided to remove it, following your suggestion.

[the remark this answers is not in the document]
Comment (ZzzzHui) [reply onchain/rollups/contracts/consensus/IConsensus.sol]: Yeah. It can be both too. I mentioned it because there were 2 events and only one of them had `/// @dev MUST be triggered`, and also 2 functions and only one of them had `/// @dev MUST trigger`

File: onchain/rollups/contracts/dapp/CartesiDApp.sol (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZzzzHui) [reply onchain/rollups/contracts/dapp/CartesiDApp.sol:193]: Yes, that's what it does. We could simply say what it is used for, because the calculation is not relevant here. But I'm ok with the current way too :)

**Comment by ZzzzHui** [comment]
> We could also make Authority inherit from EpochHashStorage directly, but then it would have to implement getEpochHash (much like AbstractConsensus)

It wouldn't. We rename the function `_getEpochHash` to `getEpochHash` and make it external.

**Comment by guidanoli** [comment]
> > We could also make Authority inherit from EpochHashStorage directly, but then it would have to implement getEpochHash (much like AbstractConsensus)
> 
> It wouldn't. We rename the function `_getEpochHash` to `getEpochHash` and make it external.

Ok, so you're suggesting to make `EpochHashStorage` an abstract contract that half-implements `IConsensus`? If we go in this direction, we could even rename it as `AbstractConsensus`. What do you think?

**Comment by ZzzzHui** [comment]
Yes. That's my suggestion

**Review by ZzzzHui — DISMISSED** [review]
Nicely done!!

**Comment by ZzzzHui** [comment]
As a side note, I think maybe we can make the future changeset files a bit less verbose (like [this one](https://github.com/cartesi/rollups-contracts/blob/5ef6bb1d9529be65d68c9881ac6650fd217f5180/onchain/rollups/.changeset/three-pigs-flow.md)), or it should be better formatted. Otherwise, it can get hard to read.

File: onchain/rollups/contracts/consensus/AbstractConsensus.sol [hunk 7]
```diff
@@ -4,12 +4,47 @@
...
+        InputRange calldata r
+    ) public view override returns (bytes32 epochHash) {
+        epochHash = _epochHashes[dapp][r.firstInputIndex][r.lastInputIndex];
+    }
+
+    /// @notice Accept a claim.
+    /// @param dapp The DApp contract address
+    /// @param r The input range
+    /// @param epochHash The epoch hash
+    /// @dev Raises an `InputRangeEmpty` error if `r` is empty.
```
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol:37]: There is no possibility of `r` not being assigned to a value. It is a parameter, which can be evaluated into a value. Of course, in Solidity, variables which are not assigned to some value explicitly are implicitly assigned to the default value of the variable type... But function arguments must be provided explicitly. Solidity doesn't support default values for parameters.

Meanwhile, the discussion of whether "empty" might be ambiguous is completely valid. Maybe if we phrase it like `represents the empty set` rather than `is empty` would make it clearer? The Wikipedia article on [intervals](https://en.wikipedia.org/wiki/Interval_(mathematics)) shows this is a common phrase:

> Each interval (a, a), [a, a), and (a, a] represents the [empty set](https://en.wikipedia.org/wiki/Empty_set), whereas [a, a] denotes the singleton set {a}. When a > b, all four notations are usually taken to represent the empty set.

**Comment by guidanoli** [comment]
> As a side note, I think maybe we can make the future changeset files a bit less verbose (like [this one](https://github.com/cartesi/rollups-contracts/blob/5ef6bb1d9529be65d68c9881ac6650fd217f5180/onchain/rollups/.changeset/three-pigs-flow.md)), or it should be better formatted. Otherwise, it can get hard to read.

I agree that multi-line changeset files may look really cluttered on their own, but I don't think they look that bad when aggregated in a changelog. See the [PR](https://github.com/cartesi/rollups-contracts/pull/147) created by the CI for reference. I believe this is how this is going to be rendered in the final CHANGELOG as well.

File: onchain/rollups/contracts/consensus/AbstractConsensus.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol]: This has been addressed by commit 6801e97.

[the remark this answers is not in the document]
Comment (ZzzzHui) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol]: > There is no possibility of r not being assigned to a value

Wouldn't it be possible to declare `r` first without assignment and pass it via function call?

[the remark this answers is not in the document]
Comment (ZzzzHui) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol]: I think "invalid" is shorter and simpler than "represents the empty set". 
"empty set" sounds like a term that some of the readers may Wikipedia it, but "invalid" is straightforward.

**Comment by ZzzzHui** [comment]
> I agree that multi-line changeset files may look really cluttered on their own, but I don't think they look that bad when aggregated in a changelog. See the https://github.com/cartesi/rollups-contracts/pull/147 created by the CI for reference. I believe this is how this is going to be rendered in the final CHANGELOG as well.

I do agree that they dont look bad at all when aggregated and look from a distance. But if reading from item to item, then it can be difficult because it's lengthy. But also I know it's gonna be helpful if users know what they are looking for and search for it, then the details would be very helpful that way.

**Review by ZzzzHui — APPROVED** [review]

File: onchain/rollups/contracts/consensus/AbstractConsensus.sol (hunk 7 above)
[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol]: I find the term "invalid" too vague.
What makes an input range invalid?
That's why I prefer to be explicit, and say "represents the empty set".

[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol]: I don't think the concept of an empty set is alien to the average `rollups-contract` user. There are far more complex concepts in our code base.

[the remark this answers is not in the document]
Comment (guidanoli) [reply onchain/rollups/contracts/consensus/AbstractConsensus.sol]: > Wouldn't it be possible to declare `r` first without assignment and pass it via function call?

Yes, but then you're making the decision to pass an unassigned variable to a function call, which is weird.

**Review by pedroargento — APPROVED** [review]
